File: src/source/token.rs
use super::document::{Position, Range};
use super::identifier::{Identifier, Keyword};
use super::number::Number;
use crate::diagnostics::DiagnosticRef;
use std::fmt;

/// A token in Brite source code is a range of text with some simple semantic meaning. When parsing
/// a source document we produce a list of tokens whose positions when added together should be the
/// full range of the document.
#[derive(Clone, Debug, PartialEq)]
pub enum Token {
    Glyph(GlyphToken),
    Identifier(IdentifierToken),
    Number(NumberToken),
    Unexpected(UnexpectedToken),
    End(EndToken),
}

/// The range covered by a token. Every token has two start positions. The “full start” position
/// and the actual start position.
///
/// The actual start position is the true start of the token. This is where the significant
/// characters in the token begin. The “full start” position includes whitespace and comments.
#[derive(Clone, PartialEq)]
pub struct TokenRange {
    /// The token’s full starting position. Including whitespace and comments.
    full_start: Position,
    /// The token’s actual range.
    range: Range,
}

impl TokenRange {
    /// Creates a new token range.
    pub fn new(full_start: Position, range: Range) -> Self {
        debug_assert!(full_start <= range.start());
        TokenRange { full_start, range }
    }

    /// Gets the full start of this token range.
    pub fn full_start(&self) -> Position {
        self.full_start
    }

    /// Gets the range for this position _not_ including the full start.
    pub fn range(&self) -> Range {
        self.range
    }

    /// Gets the end of this token range.
    pub fn end(&self) -> Position {
        self.range.end()
    }
}

impl fmt::Debug for TokenRange {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "TokenRange({}, {:?})",
            self.full_start.utf8_index(),
            self.range
        )
    }
}

/// Some sequence of characters that helps define a construct in our programming language.
#[derive(Clone, Debug, PartialEq)]
pub struct GlyphToken {
    range: TokenRange,
    glyph: Glyph,
}

/// A glyph is some symbol which is a part of Brite source code.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub enum Glyph {
    /// Any `Keyword`.
    Keyword(Keyword),
    /// `{`
    BraceLeft,
    /// `}`
    BraceRight,
    /// `,`
    Comma,
    /// `.`
    Dot,
    /// `=`
    Equals,
    /// `(`
    ParenLeft,
    /// `)`
    ParenRight,
    /// `;`
    Semicolon,
    /// `/`
    Slash,
}

impl GlyphToken {
    pub fn new(range: TokenRange, glyph: Glyph) -> Self {
        GlyphToken { range, glyph }
    }

    pub fn keyword(range: TokenRange, keyword: Keyword) -> Self {
        Self::new(range, Glyph::Keyword(keyword))
    }

    pub fn glyph(&self) -> &Glyph {
        &self.glyph
    }
}

impl Glyph {
    /// Gets the string representation of this glyph. As a static string.
    pub fn as_str(&self) -> &'static str {
        match self {
            Glyph::Keyword(keyword) => keyword.as_str(),
            Glyph::BraceLeft => "{",
            Glyph::BraceRight => "}",
            Glyph::Comma => ",",
            Glyph::Dot => ".",
            Glyph::Equals => "=",
            Glyph::ParenLeft => "(",
            Glyph::ParenRight => ")",
            Glyph::Semicolon => ";",
            Glyph::Slash => "/",
        }
    }
}

/// Any `Identifier`.
#[derive(Clone, Debug, PartialEq)]
pub struct IdentifierToken {
    range: TokenRange,
    identifier: Identifier,
}

impl IdentifierToken {
    pub fn new(range: TokenRange, identifier: Identifier) -> Self {
        IdentifierToken { range, identifier }
    }
}

/// Any `Number`. This could be a valid number or an invalid number.
#[derive(Clone, Debug, PartialEq)]
pub struct NumberToken {
    range: TokenRange,
    /// If the number is valid our result is `Ok()`. If the number is invalid our result is `Err()`
    /// with the raw string value of the number we tried to parse.
    number: Result<Number, DiagnosticRef>,
}

impl NumberToken {
    pub fn new(range: TokenRange, number: Result<Number, DiagnosticRef>) -> Self {
        NumberToken { range, number }
    }
}

/// An unexpected token we encountered while tokenizing the program.
#[derive(Clone, Debug, PartialEq)]
pub struct UnexpectedToken {
    range: TokenRange,
    unexpected: char,
}

impl UnexpectedToken {
    pub fn new(range: TokenRange, unexpected: char) -> Self {
        UnexpectedToken { range, unexpected }
    }

    pub fn unexpected(&self) -> char {
        self.unexpected
    }
}

/// The last token in the document. Once our lexer returns `EndToken` it will continue returning
/// `EndToken` forever.
#[derive(Clone, Debug, PartialEq)]
pub struct EndToken {
    range: TokenRange,
}

impl EndToken {
    pub fn new(range: TokenRange) -> Self {
        EndToken { range }
    }
}

impl Token {
    pub fn is_glyph(&self, glyph: Glyph) -> bool {
        match self {
            Token::Glyph(token) => token.glyph == glyph,
            _ => false,
        }
    }

    pub fn is_identifier(&self) -> bool {
        match self {
            Token::Identifier(_) => true,
            _ => false,
        }
    }

    pub fn is_number(&self) -> bool {
        match self {
            Token::Number(_) => true,
            _ => false,
        }
    }

    pub fn is_end(&self) -> bool {
        match self {
            Token::End(_) => true,
            _ => false,
        }
    }

    pub fn full_range(&self) -> &TokenRange {
        match self {
            Token::Glyph(t) => &t.range,
            Token::Identifier(t) => &t.range,
            Token::Number(t) => &t.range,
            Token::Unexpected(t) => &t.range,
            Token::End(t) => &t.range,
        }
    }
}

impl Into<Token> for GlyphToken {
    fn into(self) -> Token {
        Token::Glyph(self)
    }
}

impl Into<Token> for IdentifierToken {
    fn into(self) -> Token {
        Token::Identifier(self)
    }
}

impl Into<Token> for NumberToken {
    fn into(self) -> Token {
        Token::Number(self)
    }
}

impl Into<Token> for UnexpectedToken {
    fn into(self) -> Token {
        Token::Unexpected(self)
    }
}

impl Into<Token> for EndToken {
    fn into(self) -> Token {
        Token::End(self)
    }
}

File: src/Brite/Syntax/Token.hs
-- Responsible for turning a Brite source text into a format which may be parsed. This includes
-- tokenizing the document and determining positions.

{-# LANGUAGE OverloadedStrings #-}

module Brite.Syntax.Token
  ( Token(..)
  , TokenKind(..)
  , NumberToken(..)
  , unexpectedToken
  , EndToken(..)
  , endTokenRange
  , Trivia(..)
  , Newline(..)
  , Comment(..)
  , TriviaSide(..)
  , isTriviaWhitespace
  , tokenSource
  , endTokenSource
  , tokensTrimmedSource
  , removeTrailingSpaces
  , debugTokens
  ) where

import Brite.Diagnostic
import Brite.Syntax.Glyph
import Brite.Syntax.Identifier
import Brite.Syntax.Number
import Brite.Syntax.Range
import Data.Char
import Data.Text (Text)
import qualified Data.Text as Text
import qualified Data.Text.Lazy as Text.Lazy
import qualified Data.Text.Lazy.Builder as Text (Builder)
import qualified Data.Text.Lazy.Builder as Text.Builder
import qualified Data.Text.Lazy.Builder.Int as Text.Builder

-- A token is a more semantic unit for describing Brite source code documents than a character.
-- Through the tokenization of a document we add meaning by parsing low-level code elements like
-- identifiers, numbers, strings, comments, and glyphs.
--
-- * The leading trivia of a token is all of the trivia which comes before the token which has not
--   already been parsed. Comments, spaces, and newlines.
--
-- * The trailing trivia of a token is all of the trivia after a token _up until the first new
--   line_. If a token is trailed by a line comment then that is part of the trailing trivia along
--   with the newline which immediately follows but nothing else!
data Token = Token
  { tokenRange :: Range
  , tokenKind :: TokenKind
  , tokenLeadingTrivia :: [Trivia]
  , tokenTrailingTrivia :: [Trivia]
  }

-- The kind of a token.
data TokenKind
  = Glyph Glyph
  | Identifier Identifier
  | Number NumberToken
  | UnexpectedChar Char

-- The kind of a number token. When parsing a number, we might error. If we do error then we’ll use
-- the `InvalidNumberToken` variant. This way, we’ll still parse invalid numbers as numbers, but we
-- will throw at runtime.
data NumberToken
  = NumberToken Number
  | InvalidNumberToken Diagnostic Text

-- The parser ran into a token it did not recognize.
unexpectedToken :: DiagnosticMonad m => Token -> ExpectedSyntax -> m Diagnostic
unexpectedToken token expected = unexpectedSyntax (tokenRange token) unexpected expected
  where
    unexpected = case tokenKind token of
      Glyph glyph -> UnexpectedGlyph glyph
      Identifier _ -> UnexpectedIdentifier
      Number _ -> UnexpectedNumber
      UnexpectedChar c -> UnexpectedChar' c

-- The last token in a document. An end token has the position at which the document ended and all
-- the trivia between the last token and the ending.
data EndToken = EndToken
  { endTokenPosition :: Position
  , endTokenTrivia :: [Trivia]
  }

-- Gets the range covered by the end token. Starts and ends at the end token position.
endTokenRange :: EndToken -> Range
endTokenRange (EndToken { endTokenPosition = p }) = Range p p

-- Pieces of Brite syntax which (usually) don’t affect program behavior. Like comments or spaces.
data Trivia
  -- Contiguous space characters (` `).
  = Spaces Int
  -- Contiguous tab characters (`\t`). The Brite formatter prefers spaces to tabs, but since tabs
  -- are common enough we’ll add a special case for them in `Trivia`.
  | Tabs Int
  -- Contiguous newlines. The Brite formatter prefers line feeds (`\n`) to other forms
  -- of newlines.
  | Newlines Newline Int
  -- Some developer comment about the source code.
  | Comment Comment
  -- Other whitespace characters which we optimize such as obscure Unicode whitespace characters
  -- like U+00A0.
  | OtherWhitespace Char
  deriving (Show)

-- `\n`, `\r`, and `\r\n`
data Newline = LF | CR | CRLF deriving (Show)

data Comment
  -- `// ...` does not include the newline that ends the comment. Does include the `//` characters.
  = LineComment Text
  -- `/* ... */` does include the `/*` and `*/` characters.
  --
  -- If the boolean is true then we reached the end of the file before finding `*/`.
  --
  -- NOTE: Maybe we should remove support for block comments? They complicate printing and parsing
  -- and are not used for documentation. At least warn when we see a block comment?
  | BlockComment Text Bool
  deriving (Show)

-- The side on which trivia is attached to a token. Either the leading or trailing side.
data TriviaSide = Leading | Trailing

-- Is this trivia whitespace?
isTriviaWhitespace :: Trivia -> Bool
isTriviaWhitespace (Spaces _) = True
isTriviaWhitespace (Tabs _) = True
isTriviaWhitespace (Newlines _ _) = True
isTriviaWhitespace (Comment _) = False
isTriviaWhitespace (OtherWhitespace _) = True

-- Gets the source code that a token was parsed from.
tokenSource :: Token -> Text.Builder
tokenSource token =
  mconcat (map triviaSource (tokenLeadingTrivia token))
    <> tokenKindSource (tokenKind token)
    <> mconcat (map triviaSource (tokenTrailingTrivia token))

-- Gets the source code that a token kind was parsed from.
tokenKindSource :: TokenKind -> Text.Builder
tokenKindSource (Glyph g) = Text.Builder.fromText (glyphText g)
tokenKindSource (Identifier i) = Text.Builder.fromText (identifierText i)
tokenKindSource (Number (NumberToken n)) = numberSource n
tokenKindSource (Number (InvalidNumberToken _ t)) = Text.Builder.fromText t
tokenKindSource (UnexpectedChar c) = Text.Builder.singleton c

-- Gets the source code that an end token was parsed from.
endTokenSource :: EndToken -> Text.Builder
endTokenSource endToken = mconcat (map triviaSource (endTokenTrivia endToken))

-- Gets the source code that a trivia was parsed from.
triviaSource :: Trivia -> Text.Builder
triviaSource (Spaces n) = Text.Builder.fromText (Text.replicate n " ")
triviaSource (Tabs n) = Text.Builder.fromText (Text.replicate n "\t")
triviaSource (Newlines LF n) = Text.Builder.fromText (Text.replicate n "\n")
triviaSource (Newlines CR n) = Text.Builder.fromText (Text.replicate n "\r")
triviaSource (Newlines CRLF n) = Text.Builder.fromText (Text.replicate n "\r\n")
triviaSource (Comment (LineComment comment)) = Text.Builder.fromText "//" <> Text.Builder.fromText comment
triviaSource (Comment (BlockComment comment True)) = Text.Builder.fromText "/*" <> Text.Builder.fromText comment <> Text.Builder.fromText "*/"
triviaSource (Comment (BlockComment comment False)) = Text.Builder.fromText "/*" <> Text.Builder.fromText comment
triviaSource (OtherWhitespace c) = Text.Builder.singleton c

-- Builds source code from from the provided tokens removing whitespace at the beginning of the
-- code, ending of the code, and immediately before a new line (trailing whitespace).
tokensTrimmedSource :: [Token] -> Text.Builder
tokensTrimmedSource = sourceStart
  where
    trimStart t =
      let trivia = dropWhile isTriviaWhitespace (tokenLeadingTrivia t) in
        t { tokenLeadingTrivia = trivia }

    trimEnd t =
      let trivia = reverse (dropWhile isTriviaWhitespace (reverse (tokenTrailingTrivia t))) in
        t { tokenTrailingTrivia = trivia }

    sourceStart [] = source []
    sourceStart (t : ts) = source (trimStart t : ts)

    source [] = mempty
    source [t] = tokenTrimmedSource (trimEnd t)
    source (t : ts) = tokenTrimmedSource t <> source ts

-- Gets the source code that a token was parsed from. Trimming all whitespace that comes
-- immediately before a new line.
tokenTrimmedSource :: Token -> Text.Builder
tokenTrimmedSource token =
  triviaTrimmedSource (tokenLeadingTrivia token)
    <> tokenKindTrimmedSource (tokenKind token)
    <> triviaTrimmedSource (tokenTrailingTrivia token)

-- Gets the source code that a token kind was parsed from. Trimming all whitespace that comes
-- immediately before a new line.
--
-- NOTE: This function is basically the same as `tokenKindSource`. However, in the future when we
-- add string literals then we will need to trim trailing whitespace inside string literals.
tokenKindTrimmedSource :: TokenKind -> Text.Builder
tokenKindTrimmedSource (t@(Glyph _)) = tokenKindSource t
tokenKindTrimmedSource (t@(Identifier _)) = tokenKindSource t
tokenKindTrimmedSource (t@(Number _)) = tokenKindSource t
tokenKindTrimmedSource (t@(UnexpectedChar _)) = tokenKindSource t

-- Gets the source code that a list of trivia was parsed from. Trimming all whitespace that comes
-- immediately before a new line.
triviaTrimmedSource :: [Trivia] -> Text.Builder
triviaTrimmedSource = loop mempty mempty
  where
    loop source space [] = source <> space

    -- Add whitespace to the second accumulator argument.
    loop source space (t@(Spaces _) : ts) = loop source (space <> triviaSource t) ts
    loop source space (t@(Tabs _) : ts) = loop source (space <> triviaSource t) ts
    loop source space (t@(OtherWhitespace _) : ts) = loop source (space <> triviaSource t) ts

    -- Don’t add the space trivia immediately preceding a new line.
    loop source _ (t@(Newlines _ _) : ts) = loop (source <> triviaSource t) mempty ts

    -- Trim trailing whitespace from line comments.
    loop source space (Comment (LineComment comment) : ts) =
      let
        newSource =
          source <> space <> Text.Builder.fromText "//"
            <> Text.Builder.fromText (Text.dropWhileEnd isSpace comment)
      in
        loop newSource mempty ts

    -- Trim trailing whitespace from block comments.
    --
    -- We also automatically fix missing block comment endings. Even though this isn’t explicitly
    -- a part of what `triviaTrimmedSource` is described as doing.
    loop source space (Comment (BlockComment comment _) : ts) =
      let
        newSource =
          source <> space <> Text.Builder.fromText "/*"
            <> removeTrailingSpaces comment <> Text.Builder.fromText "*/"
      in
        loop newSource mempty ts

-- Removes trailing spaces which come before a new line. Trailing spaces at the end of a string are
-- left in place.
removeTrailingSpaces :: Text -> Text.Builder
removeTrailingSpaces t =
  let (t1, t2) = Text.span (\c -> c /= '\n' && c /= '\r') t in
    if Text.null t2 then
      Text.Builder.fromText t
    else
      Text.Builder.fromText (Text.dropWhileEnd isSpace t1)
        <> Text.Builder.singleton (Text.head t2)
        <> removeTrailingSpaces (Text.tail t2)

-- Debug a stream of tokens.
debugTokens :: [Token] -> EndToken -> Text.Builder
debugTokens tokens endToken = mconcat (map debugToken tokens) <> debugEndToken endToken

debugToken :: Token -> Text.Builder
debugToken (Token r k lt tt) =
  mconcat (map (debugTrivia Leading) lt)
    <> Text.Builder.fromLazyText (Text.Lazy.justifyLeft 10 ' ' (Text.Builder.toLazyText (debugRange r)))
    <> Text.Builder.fromText "| "
    <> content
    <> Text.Builder.singleton '\n'
    <> mconcat (map (debugTrivia Trailing) tt)
  where
    content = case k of
      Glyph glyph ->
        Text.Builder.fromText "Glyph `" <> Text.Builder.fromText (glyphText glyph) <> Text.Builder.singleton '`'

      Identifier identifier ->
        Text.Builder.fromText "Identifier `" <> Text.Builder.fromText (identifierText identifier) <> Text.Builder.singleton '`'

      Number (NumberToken (BinaryInteger _ _ value)) ->
        Text.Builder.fromText "BinaryInteger `" <> binary value <> Text.Builder.singleton '`'
        where
          binary i | i < 0 = binary (-i)
          binary i = go i
            where
              go 0 = Text.Builder.singleton '0'
              go 1 = Text.Builder.singleton '1'
              go n = go (n `quot` 2) <> go (n `rem` 2)

      Number (NumberToken (HexadecimalInteger _ _ value)) ->
        Text.Builder.fromText "HexadecimalInteger `" <> Text.Builder.hexadecimal value <> Text.Builder.singleton '`'

      Number (InvalidNumberToken _ raw) ->
        Text.Builder.fromText "InvalidNumber `" <> Text.Builder.fromText raw <> Text.Builder.singleton '`'

      UnexpectedChar c ->
        Text.Builder.fromText "Unexpected `" <> Text.Builder.singleton c <> Text.Builder.singleton '`'

debugEndToken :: EndToken -> Text.Builder
debugEndToken (EndToken p lt) =
  mconcat (map (debugTrivia Leading) lt)
    <> Text.Builder.fromLazyText (Text.Lazy.justifyLeft 10 ' ' (Text.Builder.toLazyText (debugPosition p)))
    <> Text.Builder.fromText "| End\n"

debugTrivia :: TriviaSide -> Trivia -> Text.Builder
debugTrivia side trivia =
  (case side of { Leading -> "+"; Trailing -> "-" })
    <> Text.Builder.fromText (Text.replicate 9 " ")
    <> Text.Builder.fromText "| "
    <> content
    <> Text.Builder.singleton '\n'
  where
    content =
      case trivia of
        Spaces n -> Text.Builder.fromText "Spaces " <> Text.Builder.decimal n
        Tabs n -> Text.Builder.fromText "Tabs " <> Text.Builder.decimal n
        Newlines LF n -> Text.Builder.fromText "Newlines LF " <> Text.Builder.decimal n
        Newlines CR n -> Text.Builder.fromText "Newlines CR " <> Text.Builder.decimal n
        Newlines CRLF n -> Text.Builder.fromText "Newlines CRLF " <> Text.Builder.decimal n
        Comment (LineComment _) -> Text.Builder.fromText "LineComment"
        Comment (BlockComment _ True) -> Text.Builder.fromText "BlockComment"
        Comment (BlockComment _ False) -> Text.Builder.fromText "BlockComment (unterminated)"
        OtherWhitespace c -> Text.Builder.fromText "OtherWhitespace `" <> Text.Builder.singleton c <> Text.Builder.singleton '`'
